store/sqlstore/post_store.go

```diff
@@ -616,10 +659,102 @@ func (s *SqlPostStore) PermanentDeleteByChannel(channelId string) error {
 	return nil
 }
 
+func (s *SqlPostStore) prepareThreadedResponse(posts []*postWithExtra, extended, reversed bool) (*model.PostList, error) {
+	list := model.NewPostList()
+	var userIds []string
+	userIdMap := map[string]bool{}
+	for _, thread := range posts {
+		for _, participantId := range thread.ThreadParticipants {
+			if _, ok := userIdMap[participantId]; !ok {
+				userIdMap[participantId] = true
+				userIds = append(userIds, participantId)
+			}
+		}
+	}
+	var users []*model.User
+	if extended {
+		var err error
+		users, err = s.User().GetProfileByIds(userIds, &store.UserGetByIdsOpts{}, true)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		for _, userId := range userIds {
+			users = append(users, &model.User{Id: userId})
+		}
+	}
+	processPost := func(p *postWithExtra) error {
+		p.Post.ReplyCount = p.ThreadReplyCount
+		for _, th := range p.ThreadParticipants {
+			var participant *model.User
+			for _, u := range users {
+				if u.Id == th {
+					participant = u
+					break
+				}
+			}
+			if participant == nil {
+				return errors.New("cannot find thread participant with id=" + th)
+			}
+			p.Post.Participants = append(p.Post.Participants, participant)
+		}
+		return nil
+	}
+
+	l := len(posts)
+	for i := range posts {
+		idx := i
+		// We need to flip the order if we selected backwards
+
+		if reversed {
+			idx = l - i - 1
+		}
+		if err := processPost(posts[idx]); err != nil {
+			return nil, err
+		}
+		list.AddPost(&posts[idx].Post)
+		list.AddOrder(posts[idx].Id)
+	}
+
+	return list, nil
+}
+
+func (s *SqlPostStore) getPostsCollapsedThreads(options model.GetPostsOptions) (*model.PostList, error) {
+	var columns []string
+	for _, c := range postSliceColumns() {
+		columns = append(columns, "Posts."+c)
+	}
+	columns = append(columns, "COALESCE(Threads.ReplyCount, 0) as ThreadReplyCount", "COALESCE(Threads.LastReplyAt, 0) as LastReplyAt", "COALESCE(Threads.Participants, '[]') as ThreadParticipants")
+	var posts []*postWithExtra
+	offset := options.PerPage * options.Page
+
+	postFetchQuery, args, _ := s.getQueryBuilder().
+		Select(columns...).
+		From("Posts").
+		LeftJoin("Threads ON Threads.PostId = Id").
+		Where(sq.Eq{"DeleteAt": 0}).
+		Where(sq.Eq{"Posts.ChannelId": options.ChannelId}).
+		Where(sq.Eq{"RootId": ""}).
+		Limit(uint64(options.PerPage)).
+		Offset(uint64(offset)).
+		OrderBy("CreateAt DESC").ToSql()
```

Comment:
If my reading of this is correct, this essentially computes the list of posts needed to fulfill a page on a paginated list. Since we are ordering by createat we are making the page computation non-deterministic in the scenario that we have multiple posts with the same exact createat i.e:

### Total dataset
```
posts
| id |     createat |
---------------------
| 1 | 1609859599058 |
| 2 | 1609859599059 |
| 3| 1609859599059 |
| 4 | 1609859599060 |
| 5 | 1609859599061 |
```

let's assume a page size of 2

### computing first page:
First ordering can return posts[1, 2] but it can also return posts[1,3]

Thoughts? Apologies if i am missing any context here.

Reply:
@nronas - first of all, thanks for taking a look, always good to have another pair of eyes on the code :)
second - yes, I think you are correct, however, @calebroseland has already brought this up and we have a pending ticket https://mattermost.atlassian.net/browse/MM-31711 to move to cursor-based paging to avoid this issue.
If you don't think that that would be a sufficient fix - please let me know

Reply:
awesome, wasn't aware of the ticket. Cursor based pagination should do the trick 👍